data-products/src/main/scala/org/sunbird/analytics/util/TextBookUtils.scala

```diff
@@ -0,0 +1,183 @@
+        tempValue = output._1
+        counterQrLinked = output._3
+        counterNotLinked = output._4
+        term1NotLinked = output._5
+        term2NotLinked = output._6
+      }
+    })
+    (counterValue,tempValue,counterQrLinked,counterNotLinked,term1NotLinked,term2NotLinked)
+  }
+
+  def generateETBTextbookReport(response: ContentInfo)(implicit sc: SparkContext): List[etbTextbookData] = {
+    var qrLinkedContent = 0
+    var qrNotLinked = 0
+    var totalLeafNodes = 0
+    var leafNodeswithoutContent = 0
+    var textBookReport = List[etbTextbookData]()
+
+    if(response!=null && response.children.size>0) {
+      val etbTextbook = parseETBTextbook(response.children.get,response,0,0,0,0)
+      qrLinkedContent = etbTextbook._1
+      qrNotLinked = etbTextbook._2+1
+      leafNodeswithoutContent = etbTextbook._3
+      totalLeafNodes = etbTextbook._4
+      val medium = if(response.medium!=null) response.medium.asInstanceOf[List[String]].mkString(",") else ""
+      val subject = if(response.subject!=null) response.subject.asInstanceOf[List[String]].mkString(",") else ""
+      val gradeLevel = if(response.gradeLevel!=null) response.gradeLevel.mkString(",") else ""
+      val textbookDf = etbTextbookData(response.channel,response.identifier,response.name,medium,gradeLevel,subject,response.status,response.createdOn.substring(0,10),response.lastUpdatedOn.substring(0,10),response.leafNodesCount,qrLinkedContent,qrNotLinked,totalLeafNodes,leafNodeswithoutContent)
+      textBookReport=textbookDf::textBookReport
+    }
+    textBookReport
+  }
+
+  def parseETBTextbook(data: List[ContentInfo], response: ContentInfo, contentLinked: Integer, contentNotLinkedQR:Integer, leafNodesContent:Integer, leafNodesCount:Integer)(implicit sc: SparkContext): (Integer,Integer,Integer,Integer) = {
+    var qrLinkedContent = contentLinked
+    var contentNotLinked = contentNotLinkedQR
+    var leafNodeswithoutContent = leafNodesContent
+    var totalLeafNodes = leafNodesCount
+
+    data.map(units => {
+      if(units.children.size==0){ totalLeafNodes=totalLeafNodes+1 }
+      if(units.children.size==0 && units.leafNodesCount==0) { leafNodeswithoutContent=leafNodeswithoutContent+1 }
+      if(units.dialcodes!=null){
+        if(units.leafNodesCount>0) { qrLinkedContent=qrLinkedContent+1 }
+        else { contentNotLinked=contentNotLinked+1 }
+      }
+
+      if(units.contentType.get=="TextBookUnit") {
+        val output = parseETBTextbook(units.children.getOrElse(List[ContentInfo]()),response,qrLinkedContent,contentNotLinked,leafNodeswithoutContent,totalLeafNodes)
+        qrLinkedContent = output._1
+        contentNotLinked = output._2
+        leafNodeswithoutContent = output._3
+        totalLeafNodes = output._4
+      }
+    })
+    (qrLinkedContent,contentNotLinked,leafNodeswithoutContent,totalLeafNodes)
+  }
+}
```

Comment:
Same with this method. A recursive function should append, but not replace values